src/utils/prim-parsing.fs

```diff
@@ -361,38 +361,49 @@ module internal Implementation =
 #endif
-                    for i = 0 to n - 1 do                                                                             
+                    // For every range to reduce merge it
+                    for i = 0 to n - 1 do
                         if valueStack.IsEmpty then failwith "empty symbol stack"
-                        let topVal = valueStack.Peep()
+                        let topVal = valueStack.Peep()                                  // Grab topVal
                         valueStack.Pop()
                         stateStack.Pop()
-                        ruleValues.[(n-i)-1] <- topVal.value  
-                        ruleStartPoss.[(n-i)-1] <- topVal.startPos  
-                        ruleEndPoss.[(n-i)-1] <- topVal.endPos  
-                        if i = 0 then lhsPos.[1] <- topVal.endPos                                     
-                        if i = n - 1 then lhsPos.[0] <- topVal.startPos
-
-                    // Use the lookahead token to populate the locations if the rhs is empty                        
-                    if n = 0 then 
+
+                        let ruleIndex = (n-i)-1
+                        ruleValues.[ruleIndex] <- topVal.value
+                        ruleStartPoss.[ruleIndex] <- topVal.startPos
+                        ruleEndPoss.[ruleIndex] <- topVal.endPos
+
+                        if i = 0 then
+                            // Initial range
+                            lhsPos.[0] <- topVal.startPos
+                            lhsPos.[1] <- topVal.endPos
+                        elif topVal.startPos.FileIndex = lhsPos.[1].FileIndex && topVal.startPos.Line <= lhsPos.[1].Line then
+                            // Reduce range if same file as the initial end point
+                            lhsPos.[0] <- topVal.startPos
+
+                    // Use the lookahead token to populate the locations if the rhs is empty
+                    if n = 0 then
                         if haveLookahead then 
-                           lhsPos.[0] <- lookaheadStartPos                                                                     
-                           lhsPos.[1] <- lookaheadEndPos                                                                       
+                           lhsPos.[0] <- lookaheadStartPos
+                           lhsPos.[1] <- lookaheadEndPos
                         else 
                            lhsPos.[0] <- lexbuf.StartPos
                            lhsPos.[1] <- lexbuf.EndPos
-                    try                                                                                               
-                          // printf "reduce %d\n" prod                                                       
-                        let redResult = reduction parseState                                                          
-                        valueStack.Push(ValueInfo(redResult, lhsPos.[0], lhsPos.[1]))
+                    try
```

Comment:
The diff could be made minimal, though it's no big deal, the other changes are an improvement

Reply:
I know, but my OCD doesn't work that way :-(